File: scripts/download-textures.js
#!/usr/bin/env node

/**
 * Texture Download Script
 *
 * Downloads CC0-licensed PBR textures from ambientCG and Poly Haven
 * for realistic building material visualization.
 *
 * Usage: node scripts/download-textures.js
 */

const https = require('https')
const fs = require('fs')
const path = require('path')

const TEXTURES_DIR = path.join(__dirname, '../public/textures')

/**
 * Manual Download Instructions
 *
 * NOTE: Automatic downloads from ambientCG require authentication.
 * Please download textures manually from the websites below and place them
 * in the appropriate folders.
 *
 * This script will create a guide to help you download the textures manually.
 */
const TEXTURE_INSTRUCTIONS = {
  wall: [
    {
      name: 'concrete',
      source: 'ambientCG',
      assetId: 'Concrete034',
      url: 'https://ambientcg.com/view?id=Concrete034',
      maps: ['Color', 'NormalGL', 'Roughness'],
      instructions: 'Download 2K-JPG version, extract Color, NormalGL, and Roughness maps',
    },
    {
      name: 'brick',
      source: 'ambientCG',
      assetId: 'Bricks074',
      url: 'https://ambientcg.com/view?id=Bricks074',
      maps: ['Color', 'NormalGL', 'Roughness'],
      instructions: 'Download 2K-JPG version, extract Color, NormalGL, and Roughness maps',
    },
    {
      name: 'wood-planks',
      source: 'ambientCG',
      assetId: 'WoodSiding013',
      url: 'https://ambientcg.com/view?id=WoodSiding013',
      maps: ['Color', 'NormalGL', 'Roughness'],
      instructions: 'Download 2K-JPG version, extract Color, NormalGL, and Roughness maps',
    },
    {
      name: 'aerated-concrete',
      source: 'ambientCG',
      assetId: 'Concrete018',
      url: 'https://ambientcg.com/view?id=Concrete018',
      maps: ['Color', 'NormalGL', 'Roughness'],
      instructions: 'Download 2K-JPG version, extract Color, NormalGL, and Roughness maps',
    },
  ],
  roof: [
    {
      name: 'roof-tiles',
      source: 'ambientCG',
      assetId: 'Tiles092',
      url: 'https://ambientcg.com/view?id=Tiles092',
      maps: ['Color', 'NormalGL', 'Roughness'],
      instructions: 'Download 2K-JPG version, extract Color, NormalGL, and Roughness maps',
    },
    {
      name: 'metal-roof',
      source: 'ambientCG',
      assetId: 'Metal032',
      url: 'https://ambientcg.com/view?id=Metal032',
      maps: ['Color', 'NormalGL', 'Roughness', 'Metalness'],
      instructions:
        'Download 2K-JPG version, extract Color, NormalGL, Roughness, and Metalness maps',
    },
  ],
  floor: [
    {
      name: 'concrete-floor',
      source: 'ambientCG',
      assetId: 'Concrete032',
      url: 'https://ambientcg.com/view?id=Concrete032',
      maps: ['Color', 'NormalGL', 'Roughness'],
      instructions: 'Download 2K-JPG version, extract Color, NormalGL, and Roughness maps',
    },
    {
      name: 'wood-floor',
      source: 'ambientCG',
      assetId: 'Wood026',
      url: 'https://ambientcg.com/view?id=Wood026',
      maps: ['Color', 'NormalGL', 'Roughness'],
      instructions: 'Download 2K-JPG version, extract Color, NormalGL, and Roughness maps',
    },
  ],
  window: [
    {
      name: 'glass',
      source: 'ambientCG',
      assetId: 'Glass001',
      url: 'https://ambientcg.com/view?id=Glass001',
      maps: ['Color', 'NormalGL', 'Roughness'],
      instructions: 'Download 2K-JPG version, extract Color, NormalGL, and Roughness maps',
    },
  ],
  door: [
    {
      name: 'wood-door',
      source: 'ambientCG',
      assetId: 'Wood049',
      url: 'https://ambientcg.com/view?id=Wood049',
      maps: ['Color', 'NormalGL', 'Roughness'],
      instructions: 'Download 2K-JPG version, extract Color, NormalGL, and Roughness maps',
    },
  ],
}

/**
 * Generate manual download guide
 */
function generateDownloadGuide() {
  console.log('📥 TEXTURE DOWNLOAD GUIDE')
  console.log('═'.repeat(70))
  console.log('\nAutomatic downloads from ambientCG require manual steps.')
  console.log('Please follow the instructions below to download textures:\n')

  let textureCount = 0

  for (const [category, textures] of Object.entries(TEXTURE_INSTRUCTIONS)) {
    console.log(`\n📁 ${category.toUpperCase()}`)
    console.log('─'.repeat(70))

    for (const texture of textures) {
      textureCount++
      console.log(`\n${textureCount}. ${texture.name.toUpperCase()}`)
      console.log(`   Source: ${texture.source}`)
      console.log(`   URL: ${texture.url}`)
      console.log(`   Steps:`)
      console.log(`   a) Visit the URL above`)
      console.log(`   b) Click "Download" button`)
      console.log(`   c) Select "2K-JPG" format`)
      console.log(`   d) Extract the downloaded ZIP file`)
      console.log(`   e) Copy these files to public/textures/${category}/:`)

      texture.maps.forEach((mapType) => {
        const mapTypeLower = mapType.toLowerCase().replace('normalgl', 'normal')
        console.log(
          `      - ${texture.assetId}_2K_${mapType}.jpg → ${texture.name}-${mapTypeLower}.jpg`
        )
      })
    }
  }

  console.log('\n' + '═'.repeat(70))
  console.log('\n💡 QUICK ALTERNATIVE: Use Placeholder Textures')
  console.log('─'.repeat(70))
  console.log('The application will work with solid colors if textures are not downloaded.')
  console.log('Textures enhance visual quality but are not required for basic functionality.\n')
}

/**
 * Check for existing textures
 */
function checkExistingTextures() {
  let existingCount = 0
  let missingCount = 0
  const missing = []

  console.log('\n🔍 CHECKING EXISTING TEXTURES')
  console.log('═'.repeat(70))

  for (const [category, textures] of Object.entries(TEXTURE_INSTRUCTIONS)) {
    for (const texture of textures) {
      for (const mapType of texture.maps) {
        const mapTypeLower = mapType.toLowerCase().replace('normalgl', 'normal')
        const filename = `${texture.name}-${mapTypeLower}.jpg`
        const filepath = path.join(TEXTURES_DIR, category, filename)

        if (fs.existsSync(filepath)) {
          existingCount++
        } else {
          missingCount++
          missing.push(`${category}/${filename}`)
        }
      }
    }
  }

  console.log(`\n✅ Found: ${existingCount} texture files`)
  console.log(`❌ Missing: ${missingCount} texture files`)

  if (missingCount > 0 && existingCount === 0) {
    console.log('\n📋 All textures need to be downloaded.')
  } else if (missingCount > 0) {
    console.log('\n📋 Missing textures:')
    missing.slice(0, 10).forEach((file) => console.log(`   - ${file}`))
    if (missing.length > 10) {
      console.log(`   ... and ${missing.length - 10} more`)
    }
  } else {
    console.log('\n🎉 All textures are present!')
  }

  console.log('\n' + '═'.repeat(70))
  return { existingCount, missingCount }
}

/**
 * Create README in textures directory
 */
function createReadme() {
  const readmePath = path.join(TEXTURES_DIR, 'README.md')
  const content = `# Textures

This directory contains CC0-licensed PBR textures for building visualization.

## Sources

All textures are sourced from:
- **ambientCG** (https://ambientcg.com/) - CC0 License
- Resolution: 2K (2048x2048)
- Format: JPEG

## License

All textures in this directory are released under CC0 (Creative Commons Zero).
This means they are free to use for any purpose, including commercial projects,
without attribution (though attribution is appreciated).

## Maps Included

Each material includes:
- **Color Map** (Albedo/Diffuse)
- **Normal Map** (Surface details)
- **Roughness Map** (Surface smoothness)
- **Metalness Map** (where applicable)

## Categories

- \`wall/\` - Wall materials (concrete, brick, wood, aerated concrete)
- \`roof/\` - Roof materials (tiles, metal)
- \`floor/\` - Floor materials (concrete, wood)
- \`window/\` - Glass materials
- \`door/\` - Door materials (wood)

## Usage

Textures are automatically loaded by the material system.
See \`src/lib/constants/materials.ts\` for material definitions.

## Re-downloading

To re-download textures, delete the desired files and run:
\`\`\`bash
node scripts/download-textures.js
\`\`\`

## Adding Custom Textures

You can add your own textures by placing them in the appropriate category folder.
Follow the naming convention: \`{material-name}-{map-type}.jpg\`
`

  fs.writeFileSync(readmePath, content)
  console.log('\n📄 Created README.md in textures directory')
}

// Run the script
;(async () => {
  try {
    const stats = checkExistingTextures()

    if (stats.missingCount > 0) {
      generateDownloadGuide()
    }

    createReadme()

    console.log('\n📚 For detailed information, see:')
    console.log('   - TEXTURES.md (complete documentation)')
    console.log('   - public/textures/README.md (texture inventory)\n')
  } catch (error) {
    console.error('\n❌ Error:', error.message)
    process.exit(1)
  }
})()

File: scripts/download-textures.py
#!/usr/bin/env python3
"""
Automatic Texture Downloader for 3D Building Viewer

Downloads CC0-licensed PBR textures from direct CDN links,
extracts them, and organizes them into the correct directory structure.

Usage: python3 scripts/download-textures.py
"""

import os
import sys
import zipfile
import urllib.request
import urllib.error
from pathlib import Path
from typing import Dict, List, Tuple

# Base directories
SCRIPT_DIR = Path(__file__).parent
PROJECT_DIR = SCRIPT_DIR.parent
TEXTURES_DIR = PROJECT_DIR / 'public' / 'textures'

# Direct download URLs from ambientCG CDN
# Format: https://ambientcg.com/get?file=AssetID_Resolution-Format.zip
BASE_URL = "https://ambientcg.com/get?file="

# Material definitions with ambientCG Asset IDs
MATERIALS = {
    'wall': [
        {'name': 'concrete', 'asset_id': 'Concrete034', 'maps': ['Color', 'NormalGL', 'Roughness']},
        {'name': 'brick', 'asset_id': 'Bricks074', 'maps': ['Color', 'NormalGL', 'Roughness']},
        {'name': 'wood-planks', 'asset_id': 'WoodSiding013', 'maps': ['Color', 'NormalGL', 'Roughness']},
        {'name': 'aerated-concrete', 'asset_id': 'Concrete018', 'maps': ['Color', 'NormalGL', 'Roughness']},
    ],
    'roof': [
        {'name': 'roof-tiles', 'asset_id': 'Tiles092', 'maps': ['Color', 'NormalGL', 'Roughness']},
        {'name': 'metal-roof', 'asset_id': 'Metal032', 'maps': ['Color', 'NormalGL', 'Roughness', 'Metalness']},
    ],
    'floor': [
        {'name': 'concrete-floor', 'asset_id': 'Concrete032', 'maps': ['Color', 'NormalGL', 'Roughness']},
        {'name': 'wood-floor', 'asset_id': 'Wood026', 'maps': ['Color', 'NormalGL', 'Roughness']},
    ],
    'window': [
        {'name': 'glass', 'asset_id': 'Glass001', 'maps': ['Color', 'NormalGL', 'Roughness']},
    ],
    'door': [
        {'name': 'wood-door', 'asset_id': 'Wood049', 'maps': ['Color', 'NormalGL', 'Roughness']},
    ],
}


def print_header():
    """Print script header"""
    print("🎨 ambientCG Texture Downloader")
    print("=" * 70)
    print("Source: ambientCG (CC0 License)")
    print("Resolution: 2K")
    print("Format: JPG")
    print("=" * 70)
    print()


def create_directories():
    """Create necessary directory structure"""
    for category in MATERIALS.keys():
        category_dir = TEXTURES_DIR / category
        category_dir.mkdir(parents=True, exist_ok=True)
    print("✅ Directory structure created\n")


def download_file(url: str, dest: Path) -> bool:
    """Download a file from URL to destination with progress"""
    try:
        # Create request with proper headers
        req = urllib.request.Request(url, headers={
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
            'Accept': 'application/zip,*/*'
        })
        
        with urllib.request.urlopen(req, timeout=120) as response:
            # Check if we got a valid response
            content_type = response.headers.get('Content-Type', '')
            if 'text/html' in content_type:
                print(f"   ❌ Got HTML instead of ZIP (asset may not exist)")
                return False
            
            total_size = int(response.headers.get('Content-Length', 0))
            
            with open(dest, 'wb') as f:
                downloaded = 0
                chunk_size = 32768  # 32KB chunks
                
                while True:
                    chunk = response.read(chunk_size)
                    if not chunk:
                        break
                    
                    f.write(chunk)
                    downloaded += len(chunk)
                    
                    if total_size > 0:
                        progress = (downloaded / total_size) * 100
                        mb_downloaded = downloaded / (1024 * 1024)
                        mb_total = total_size / (1024 * 1024)
                        print(f"\r   Progress: {progress:.1f}% ({mb_downloaded:.1f}MB / {mb_total:.1f}MB)", 
                              end='', flush=True)
                
                if total_size > 0:
                    print()  # New line after progress
        
        # Verify the file was downloaded and has content
        if dest.stat().st_size == 0:
            print(f"   ❌ Downloaded file is empty")
            dest.unlink()
            return False
            
        return True
    except urllib.error.HTTPError as e:
        print(f"\n   ❌ HTTP Error {e.code}: {e.reason}")
        return False
    except urllib.error.URLError as e:
        print(f"\n   ❌ Connection Error: {e.reason}")
        return False
    except Exception as e:
        print(f"\n   ❌ Error: {e}")
        return False


def extract_and_rename(zip_path: Path, category: str, material: Dict) -> Tuple[int, int]:
    """Extract ZIP and rename files to expected format"""
    extracted = 0
    skipped = 0
    
    try:
        # First, list files in ZIP
        with zipfile.ZipFile(zip_path, 'r') as zip_ref:
            file_list = zip_ref.namelist()
            print(f"   📦 ZIP contains {len(file_list)} files")
            
            for map_type in material['maps']:
                # Find files matching this map type
                map_type_lower = map_type.lower().replace('normalgl', 'normal')
                
                # Try different possible naming patterns
                possible_patterns = [
                    f"{material['asset_id']}_2K_{map_type}.jpg",
                    f"{material['asset_id']}_2K-JPG_{map_type}.jpg",
                    f"{material['asset_id']}_2K_JPG_{map_type}.jpg",
                ]
                
                matching_file = None
                for pattern in possible_patterns:
                    matches = [f for f in file_list if pattern in f or f.endswith(pattern)]
                    if matches:
                        matching_file = matches[0]
                        break
                
                # If still not found, try case-insensitive search
                if not matching_file:
                    for f in file_list:
                        if map_type.lower() in f.lower() and f.endswith('.jpg'):
                            matching_file = f
                            break
                
                if matching_file:
                    target_filename = f"{material['name']}-{map_type_lower}.jpg"
                    target_path = TEXTURES_DIR / category / target_filename
                    
                    # Check if already exists
                    if target_path.exists() and target_path.stat().st_size > 0:
                        print(f"   ⏭️  {map_type_lower}: already exists")
                        skipped += 1
                        continue
                    
                    # Extract and save
                    with zip_ref.open(matching_file) as source:
                        with open(target_path, 'wb') as target:
                            target.write(source.read())
                    
                    size_mb = target_path.stat().st_size / (1024 * 1024)
                    print(f"   ✅ {map_type_lower}: extracted ({size_mb:.1f}MB)")
                    extracted += 1
                else:
                    print(f"   ⚠️  {map_type_lower}: not found in ZIP")
                    # Show first few files for debugging
                    if len(file_list) > 0:
                        print(f"      Available: {', '.join(file_list[:3])}")
    
    except zipfile.BadZipFile:
        print(f"   ❌ Invalid or corrupt ZIP file")
    except Exception as e:
        print(f"   ❌ Extraction error: {e}")
    
    return extracted, skipped


def download_material(category: str, material: Dict) -> Tuple[int, int]:
    """Download and process a single material"""
    print(f"\n📦 {material['name'].upper()} ({material['asset_id']})")
    
    # Check if all files already exist
    all_exist = True
    for map_type in material['maps']:
        map_type_lower = map_type.lower().replace('normalgl', 'normal')
        target_path = TEXTURES_DIR / category / f"{material['name']}-{map_type_lower}.jpg"
        if not target_path.exists() or target_path.stat().st_size == 0:
            all_exist = False
            break
    
    if all_exist:
        print(f"   ✅ All textures already present")
        return 0, len(material['maps'])
    
    # Construct download URL
    filename = f"{material['asset_id']}_2K-JPG.zip"
    download_url = f"{BASE_URL}{filename}"
    temp_zip = TEXTURES_DIR / f"temp_{material['asset_id']}.zip"
    
    print(f"   ⬇️  Downloading from ambientCG...")
    print(f"   URL: {download_url}")
    
    if download_file(download_url, temp_zip):
        # Extract and rename
        extracted, skipped = extract_and_rename(temp_zip, category, material)
        
        # Clean up ZIP
        try:
            temp_zip.unlink()
        except:
            pass
        
        return extracted, skipped
    else:
        return 0, 0


def main():
    """Main execution function"""
    print_header()
    
    # Create directories
    create_directories()
    
    # Statistics
    total_downloaded = 0
    total_skipped = 0
    total_failed = 0
    
    # Process each category
    for category, materials in MATERIALS.items():
        print(f"\n📁 {category.upper()}")
        print("─" * 70)
        
        for material in materials:
            extracted, skipped = download_material(category, material)
            total_downloaded += extracted
            total_skipped += skipped
            
            if extracted == 0 and skipped == 0:
                total_failed += len(material['maps'])
    
    # Summary
    print("\n" + "=" * 70)
    print("\n✨ Download Complete!")
    print(f"   ✅ Downloaded: {total_downloaded} texture files")
    print(f"   ⏭️  Skipped: {total_skipped} files (already exist)")
    
    if total_failed > 0:
        print(f"   ❌ Failed: {total_failed} files")
        print("\n💡 Note: Some assets may have been renamed or removed from ambientCG.")
        print("   Visit https://ambientcg.com/ to find alternative textures.")
    
    print("\n📝 All textures are CC0-licensed from ambientCG")
    print("   https://ambientcg.com/")
    
    if total_downloaded > 0:
        print("\n🎉 Textures are ready! Start the app with: npm run dev")
    print()


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\n⚠️  Download interrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n❌ Unexpected error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)
